### PR TITLE
settings: Add option to mark read on scroll only in conversation views

### PR DESCRIPTION
**Pull request by chrisbobbe**
Thanks to the handy new `InputRowRadioButtons` component we added in edf53220f, this is now a quick user-visible feature we can check off the list.

Screenshots below.

Fixes: #5241
Replaces: #5294

### CODE REVIEW

**Comment by chrisbobbe** [comment]
Before:



After:

**Comment by chrisbobbe** [comment]
Ah, looking at the settings screen there, the right-facing arrows aren't lined up. Let's check back in after #5446 is merged; I expect that to go in the right direction towards fixing that (see screenshots there), and may even fix it entirely. Marking as blocked on #5446.

File: src/common/InputRowRadioButtons.js [hunk 1]
```diff
@@ -164,7 +170,7 @@ export default function InputRowRadioButtons<TItemKey: string>(props: Props<TIte
           <ZulipTextIntl text={selectedItem.title} style={styles.valueTitle} />
         </View>
         <View style={styles.iconRightWrapper}>
-          <IconRight size={kRightArrowIconSize} color={BRAND_COLOR} />
+          <IconRight size={kRightArrowIconSize} color={themeData.color} />
```
Comment (gnprice) [inline src/common/InputRowRadioButtons.js:173]: What's the effect of this first commit -- is there an existing screen where it changes something?

Comment (chrisbobbe) [reply src/common/InputRowRadioButtons.js:173]: Yep, the edit-stream screen. Here's a before-and-after:

Comment (gnprice) [reply src/common/InputRowRadioButtons.js:173]: Great, thanks. Seems good.

**Comment by gnprice** [comment]
#5446 has been merged, so I think the next step is to rebase atop that and see how that alignment looks.

**Comment by chrisbobbe** [comment]
> #5446 has been merged, so I think the next step is to rebase atop that and see how that alignment looks.

Yep, I agree. I haven't done that yet in this latest revision—for now, this just resolves rebase conflicts, and adds two commits on top that convert to using Flow enums now that #5444 has landed. We can see how we like that for this use case. 🙂 For now I have to run, though.

**Comment by gnprice** [comment]
Thanks for the revision! Will take a look at the code momentarily.

@alya, do you have any thoughts on the text in the new screen? See the lower-right screenshot in the first comment above.

File: src/reduxTypes.js [hunk 2]
```diff
@@ -375,6 +375,12 @@ export type PerAccountSettingsState = $ReadOnly<{
   ...
 }>;
 
+export enum MarkMessagesReadOnScroll {
+  Always = 0,
+  Never = 1,
+  ConversationViewsOnly = 3,
```
Comment (gnprice) [inline src/reduxTypes.js:381]: Hmm -- why skip 2? Do the values end up mattering?

File: src/storage/__tests__/migrations-test.js [hunk 3]
```diff
@@ -84,10 +85,27 @@ describe('migrations', () => {
     },
   };
 
+  // What `base` becomes after migrations up through 52.
+  const base52 = {
+    migrations: { version: 52 },
+    ...base37,
```
Comment (gnprice) [inline src/storage/__tests__/migrations-test.js:91]: This will clobber `migrations`, no?

File: src/settings/SettingsScreen.js [hunk 4]
```diff
@@ -47,11 +49,27 @@ export default function SettingsScreen(props: Props): Node {
...
+            title: 'Only in conversation views',
+            subtitle:
+              'Messages will be marked as read in single-topic or private-message views, but not in interleaved views, such as whole-stream views.',
+          },
+        ]}
         onValueChange={value => {
-          dispatch(setGlobalSettings({ doNotMarkMessagesAsRead: value }));
+          dispatch(
+            setGlobalSettings({
+              markMessagesReadOnScroll: MarkMessagesReadOnScroll.cast(value),
```
Comment (gnprice) [inline src/settings/SettingsScreen.js:70]: Hmm, awkward to have to do this. (Also the static casts `(…: number)` above.)

It looks like the root of this is that we say `key={item.key}` in `SelectableOptionsScreen`, passing these keys as React `key` pseudoprops. Ideally the type accepted there could be like "string, or number, or any enum of string or number", but there's no way to express the "any enum" part:
https://flow.org/en/docs/enums/using-enums/#toc-generic-enums
so it isn't.

I think in light of this bit I'd lean toward skipping the enum here, and sticking with the disjoint union of strings. There isn't a lot that we gain in this instance from the enum. But this price for using an enum isn't too bad, either.

File: src/storage/__tests__/migrations-test.js (hunk 3 above)
Comment (chrisbobbe) [reply src/storage/__tests__/migrations-test.js:91]: Oops, I should switch the order of these two lines.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by gnprice** [comment]
Thanks! Looks good -- merging.

@alya, if you have any thoughts on the settings text, happy to make a change.

**Comment by alya** [comment]
Sorry I missed the question earlier! I started a thread about tweaking the hint on CZO: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/mark.20as.20read.20setting.20.23M5469/near/1433721

**Pull request merged**